packages/flutter/lib/src/material/theme_data.dart

```diff
@@ -1634,3 +1671,152 @@ class _FifoCache<K, V> {
     return _cache[key] = loader();
   }
 }
+
+/// Represents the user interface visual density.
+///
+/// Density, in the context of a UI, is the vertical and horizontal
+/// "compactness" of the elements in the UI. It is unitless, since it means
+/// different things to different UI elements.
+///
+/// For example, for buttons, it affects the spacing around the child of the
+/// button. For lists, it affects the distance between baselines of entries in
+/// the list. For chips, it only affects the vertical size, not the horizontal
+/// size.
+class VisualDensity extends Diagnosticable {
+  /// A const constructor for [VisualDensity].
+  ///
+  /// All of the arguments must be non-null, and [horizontal] and [vertical]
+  /// must be in the interval between [minimumDensity] and [maximumDensity],
+  /// inclusive.
+  const VisualDensity({
+    this.horizontal = 0.0,
+    this.vertical = 0.0,
+    this.interval = 4.0,
+  })  : assert(horizontal != null),
+        assert(vertical != null),
+        assert(interval != null),
+        assert(vertical <= maximumDensity),
+        assert(vertical >= minimumDensity),
+        assert(horizontal <= maximumDensity),
+        assert(horizontal >= minimumDensity);
+
+
+  /// The minimum allowed density.
+  static const double minimumDensity = -4.0;
```

Comment:
Shouldn't this be 
```dart
static const VisualDensity minimum = VisualDensity(horizontal: -4, vertical: -4);
```

The assertions could compare against minimum.vertical, etc

Reply:
Nope: "compile time expression depends on itself".